packages/plugin-ext/src/main/browser/editor/untitled-resource.ts

```diff
@@ -14,39 +14,115 @@
+
+    async resolve(uri: URI): Promise<UntitledResource> {
+        if (uri.scheme !== Schemes.UNTITLED) {
+            throw new Error('The given uri is not untitled file uri: ' + uri);
+        } else {
+            const untitledResource = this.resources.get(uri.toString());
+            if (!untitledResource) {
+                return this.createUntitledResource(this.fileResourceResolver, '', '', uri);
+            } else {
+                return untitledResource;
+            }
         }
-        throw new Error(`scheme ${uri.scheme} is not '${Schemes.UNTITLED}'`);
+    }
+
+    async  createUntitledResource(fileResourceResolver: FileResourceResolver, content?: string, language?: string, uri?: URI): Promise<UntitledResource> {
+        let extension;
+        if (language) {
+            for (const lang of monaco.languages.getLanguages()) {
+                if (lang.id === language) {
+                    if (lang.extensions) {
+                        extension = lang.extensions[0];
+                        break;
+                    }
+                }
+            }
+        }
+        return new UntitledResource(this.resources, uri ? uri : new URI().withScheme(Schemes.UNTITLED).withPath(`/Untitled-${index++}${extension ? extension : ''}`),
+            fileResourceResolver, content);
     }
 }
 
 export class UntitledResource implements Resource {
 
-    constructor(public uri: URI, private content?: string) {
-        resources.set(this.uri.toString(), this);
-    }
+    private fileResource?: FileResource;
+
+    protected readonly onDidChangeContentsEmitter = new Emitter<void>();
+    readonly onDidChangeContents: Event<void> = this.onDidChangeContentsEmitter.event;
 
-    readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
-        return Promise.resolve(this.content ? this.content : '');
+    constructor(private resources: Map<string, UntitledResource>, public uri: URI, private fileResourceResolver: FileResourceResolver, private content?: string) {
+        this.resources.set(this.uri.toString(), this);
     }
 
     dispose(): void {
-        resources.delete(this.uri.toString());
+        this.resources.delete(this.uri.toString());
+    }
+
+    async readContents(options?: { encoding?: string | undefined; } | undefined): Promise<string> {
+        if (this.fileResource) {
+            return this.fileResource.readContents(options);
+        } else if (this.content) {
+            return this.content;
+        } else {
+            return '';
+        }
+    }
+
+    async saveContents(content: string, options?: { encoding?: string, overwriteEncoding?: string }): Promise<void> {
+        if (!this.fileResource) {
+            this.fileResource = await this.fileResourceResolver.resolve(new URI(this.uri.path.toString()));
+        }
+        if (this.fileResource) {
+            await this.fileResource.saveContents(content, options);
+            this.fireDidChangeContents();
```

Comment:
i don't think it is required, `FileResource` knows better

Reply:
done